caikit/core/toolkit/destroyable.py

```diff
@@ -0,0 +1,54 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Common interface for destroyable threads and processes
+"""
+
+# Standard
+from typing import Any
+import abc
+
+
+class Destroyable(abc.ABC):
+    __doc__ = __doc__
+
+    @property
+    @abc.abstractmethod
+    def destroyed(self) -> bool:
+        """Return True if destroy was called, regardless of whether the
+        destroyable was alive at the time
+        """
+
+    @property
+    @abc.abstractmethod
+    def canceled(self) -> bool:
```

Comment:
This was actually my way of getting around the need to expose `events` like you had in your train executors. The crux of that issue was that the function that would call `destroy` couldn't differentiate between being called because the operation had terminated naturally, or because the user had hung up. As a result, `destroy` would be called in either case and the only way to tell the difference was to see if the `cancel_event` was set.

Having `canceled` vs `destroyed` allows us to differentiate this by interface rather than implementation and therefore allows us to not have these events have shared ownership (in fact, events are no longer needed at all).